File: backend/app/routes/simulate.py
"""Routes for simulation execution."""

from fastapi import APIRouter, HTTPException
from app.models import SimulationRequest, SimulationResponse
from app.services.simulator import validate_simulation_input, run_simulation

router = APIRouter(prefix="/api/simulate", tags=["simulation"])


@router.post("/")
async def execute_simulation(request: SimulationRequest) -> SimulationResponse:
    """
    Run a simulation on an architecture.
    
    Args:
        request: Simulation request containing nodes, edges, and steps
        
    Returns:
        SimulationResponse with the evolution of node metrics over time
    """
    try:
        # Validate input
        is_valid, error_msg = validate_simulation_input(request.nodes, request.edges)
        if not is_valid:
            raise HTTPException(status_code=400, detail=error_msg)
        
        # Validate steps parameter
        if not isinstance(request.steps, int) or request.steps < 1 or request.steps > 1000:
            raise HTTPException(status_code=400, detail="Steps must be an integer between 1 and 1000")
        
        # Run simulation
        results = run_simulation(request.nodes, request.edges, request.steps)
        
        return SimulationResponse(
            success=True,
            steps=request.steps,
            simulation=results
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Simulation failed: {str(e)}")

File: backend/src/routes/simulate.js
// Routes for simulation operations
const express = require('express');
const router = express.Router();
const { runSimulation, validateSimulationInput } = require('../services/simulator');

/**
 * POST /api/simulate
 * Run a simulation on an architecture
 * 
 * Request body:
 * {
 *   nodes: Array<Node>,
 *   edges: Array<Edge>,
 *   steps: Number (optional, default 10)
 * }
 */
router.post('/', (req, res) => {
  try {
    const { nodes, edges, steps = 10 } = req.body;

    // Validate input
    const validation = validateSimulationInput({ nodes, edges });
    if (!validation.isValid) {
      return res.status(400).json({ error: validation.error });
    }

    // Validate steps parameter
    if (!Number.isInteger(steps) || steps < 1 || steps > 1000) {
      return res.status(400).json({ error: 'Steps must be an integer between 1 and 1000' });
    }

    // Run simulation
    const results = runSimulation(nodes, edges, steps);
    
    res.json({
      success: true,
      steps: steps,
      simulation: results
    });
  } catch (error) {
    res.status(500).json({ error: 'Simulation failed: ' + error.message });
  }
});

module.exports = router;
